### ISSUE TITLE
SIGSEGV on FreeBSD with version 1.13.0

### ISSUE THREAD

**Issue by kapfenho** [NLnetLabs/unbound#376]
With version 1.13.0 I got a segmentation violation (sorry, no core file). That's the first time, after years of stability.

Last messages:
```
Dec 14 01:03:59 <daemon.notice> homer unbound[72482]: [72482:0] notice: sendto failed: Permission denied  
Dec 14 01:03:59 <kern.info> homer kernel: pid 72482 (unbound), jid 0, uid 59: exited on signal 11  
```

System:
```
FreeBSD homer 12.2-RELEASE-p1 FreeBSD 12.2-RELEASE-p1 GENERIC  amd64
```

Unbound config (using IPv4 & IPv6):
```
server:
        verbosity: 1
        statistics-interval: 3600
        interface: ::1
        interface: ...
        interface-automatic: yes
        outgoing-interface: ...
        access-control: ... allow
        root-hints: "/usr/local/etc/unbound/root.hints"
        hide-identity: yes
        hide-version: yes
        harden-algo-downgrade: yes
        auto-trust-anchor-file: "/usr/local/etc/unbound/root.key"
remote-control:
        control-enable: yes
        control-interface: 127.0.0.1
```

**Comment by rozhuk-im** [comment]
Crash for me too, build from sources: TFOCL, TFOSE, THREADS options is set.

**Comment by kapfenho** [comment]
Options here: ECDSA GOST LIBEVENT THREADS

**Comment by emikulic** [comment]
Some debugging and backtraces in https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=251821

**Comment by delphij** [comment]
I think this was caused by commit 15e8f5c6d4f85c3858c7ad8927ee50f2da8a8a54 .

The problem is that the logging code really expects a valid `addr` pointer, and now in `udp_connect` case it's no longer true.

I think this would stop unbound from crashing for the FreeBSD case:

```
diff --git a/services/outside_network.c b/services/outside_network.c
index 0886907f..e2843573 100644
--- a/services/outside_network.c
+++ b/services/outside_network.c
@@ -1901,6 +1901,9 @@ randomize_and_send_udp(struct pending* pend, sldns_buffer* packet, int timeout)
 	/* send it over the commlink */
 	if(outnet->udp_connect) {
 		if(!comm_point_send_udp_msg(pend->pc->cp, packet, NULL, 0)) {
+			log_addr(VERB_OPS, "remote address is",
+				(struct sockaddr_storage*)&pend->addr,
+				pend->addrlen);
 			portcomm_loweruse(outnet, pend->pc);
 			return 0;
 		}
diff --git a/util/netevent.c b/util/netevent.c
index 7c6da50b..27e94438 100644
--- a/util/netevent.c
+++ b/util/netevent.c
@@ -367,9 +367,14 @@ comm_point_send_udp_msg(struct comm_point *c, sldns_buffer* packet,
 #endif
 			int e;
 			fd_set_block(c->fd);
-			sent = sendto(c->fd, (void*)sldns_buffer_begin(packet), 
-				sldns_buffer_remaining(packet), 0,
-				addr, addrlen);
+			if (addr) {
+				sent = sendto(c->fd, (void*)sldns_buffer_begin(packet),
+					sldns_buffer_remaining(packet), 0,
+					addr, addrlen);
+			} else {
+				sent = send(c->fd, (void*)sldns_buffer_begin(packet),
+					sldns_buffer_remaining(packet), 0);
+			}
 			e = errno;
 			fd_set_nonblock(c->fd);
 			errno = e;
@@ -378,9 +383,13 @@ comm_point_send_udp_msg(struct comm_point *c, sldns_buffer* packet,
 	if(sent == -1) {
 		if(!udp_send_errno_needs_log(addr, addrlen))
 			return 0;
-		verbose(VERB_OPS, "sendto failed: %s", sock_strerror(errno));
-		log_addr(VERB_OPS, "remote address is", 
-			(struct sockaddr_storage*)addr, addrlen);
+		if (addr) {
+			verbose(VERB_OPS, "sendto failed: %s", sock_strerror(errno));
+			log_addr(VERB_OPS, "remote address is",
+				(struct sockaddr_storage*)addr, addrlen);
+		} else {
+			verbose(VERB_OPS, "send failed: %s", sock_strerror(errno));
+		}
 		return 0;
 	} else if((size_t)sent != sldns_buffer_remaining(packet)) {
 		log_err("sent %d in place of %d bytes", 
```

but it would introduce a layer violation; and arguably it's not safe to call `udp_send_errno_needs_log` with NULL either, although it would not cause a crash with `ENOBUFS`.  It's late here so I'm posting the patch here in the hope that someone familiar with the code can create a better fix.

**Comment by gthess** [comment]
@delphij You found a bug for sure, thanks for the patch!
I modified it and included it in https://github.com/NLnetLabs/unbound/commit/08968baec1122a58bb90d8f97ad948a75f8a5d69.

It seems it solves the problems reported here but not sure about the one reported over at https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=251821. It seems to be TCP related.

**Comment by Sp1l** [comment]
Added to FreeBSD ports in 558269 and 558271 (2020Q4)
See https://github.com/freebsd/freebsd-ports/commit/1d6fa91ce51a6a5143083401523b07eb2520baef
https://github.com/freebsd/freebsd-ports/commit/aa6db09c7b1627dab20dbe2c37f491a75268bbf3

**Comment by delphij** [comment]
> @delphij You found a bug for sure, thanks for the patch!
> I modified it and included it in [08968ba](https://github.com/NLnetLabs/unbound/commit/08968baec1122a58bb90d8f97ad948a75f8a5d69).
> 
> It seems it solves the problems reported here but not sure about the one reported over at https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=251821. It seems to be TCP related.

That one looks like a different bug.  Unfortunately I was unable to trigger it locally...

**Comment by cschuber** [comment]
> > @delphij You found a bug for sure, thanks for the patch!
> > I modified it and included it in [08968ba](https://github.com/NLnetLabs/unbound/commit/08968baec1122a58bb90d8f97ad948a75f8a5d69).
> > It seems it solves the problems reported here but not sure about the one reported over at https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=251821. It seems to be TCP related.
> 
> That one looks like a different bug. Unfortunately I was unable to trigger it locally...

I'm not able to reproduce this here either.

**Comment by wcawijngaards** [comment]
The commit may fix the tcp related issue that the others seem unable to trigger.  I could not trigger it also, but I added code to stop double additions and double deletions from the event map.  I think that may stop the failure from the libevent library.

**Comment by wcawijngaards** [comment]
The fix for #393 has been tested to fix the issue that was still pending for the issue list here.  That fixes the problems reported here, so I close the issue.  If problems persist, feel free to open a new issue, or continue to comment here.
